File: src/main/java/fp/yeyu/tos/TotemOfShadowItem.java
package fp.yeyu.tos;

import net.minecraft.entity.SpawnReason;
import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Rarity;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class TotemOfShadowItem extends Item {

    private static final Logger LOGGER = LogManager.getLogger();

    public TotemOfShadowItem() {
        super(new Item.Settings()
                .rarity(Rarity.EPIC)
                .group(ItemGroup.MISC)
                .maxCount(1)
        );
    }

    @Override
    public ActionResult useOnBlock(ItemUsageContext context) {
        if (!(context.getWorld() instanceof ServerWorld)) return super.useOnBlock(context);
        if (context.getPlayer() == null) return super.useOnBlock(context);
        final ShadowEntity shadowEntity = TOSEntry.SHADOW_ENTITY.spawn((ServerWorld) context.getWorld(), null, null, null, context.getBlockPos().add(0, 1, 0), SpawnReason.EVENT, false, false);

        if (shadowEntity == null) {
            LOGGER.error("Cannot spawn Shadow mob.");
            return super.useOnBlock(context);
        }
        context.getStack().decrement(1);
        shadowEntity.sendSpawnPacket(context.getPlayer().getUuid());
        shadowEntity.copyingEntity = context.getPlayer();
        if (!context.getWorld().isClient) shadowEntity.playSpawnParticles();
        return ActionResult.CONSUME;
    }
}

File: src/main/kotlin/fp/yeyu/tos/item/TotemOfShadowItem.kt
package fp.yeyu.tos.item

import fp.yeyu.tos.TotemOfShadowEntry
import fp.yeyu.tos.entity.ShadowEntity
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.entity.SpawnReason
import net.minecraft.item.Item
import net.minecraft.item.ItemGroup
import net.minecraft.item.ItemStack
import net.minecraft.item.ItemUsageContext
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.ActionResult
import net.minecraft.util.Rarity
import org.apache.logging.log4j.LogManager

object TotemOfShadowItem : Item(Settings()
        .rarity(Rarity.RARE)
        .group(ItemGroup.MISC)
        .maxDamage(4)
) {
    private val LOGGER = LogManager.getLogger()

    override fun useOnBlock(context: ItemUsageContext): ActionResult {
        if (context.world !is ServerWorld) return super.useOnBlock(context)
        val player = context.player ?: return super.useOnBlock(context)
        val shadowEntity: ShadowEntity = TotemOfShadowEntry.shadowEntity.spawn(
                context.world as ServerWorld,
                null,
                null,
                null,
                context.blockPos.add(0, 1, 0),
                SpawnReason.EVENT,
                false,
                false
        ) ?: run {
            LOGGER.error("Cannot spawn Shadow mob.")
            return super.useOnBlock(context)
        }

        context.stack.decrement(1)
        shadowEntity.sendSpawnPacket(player.uuid)
        shadowEntity.copyingEntity = player
        shadowEntity.playSpawnParticles()
        EnchantmentHelper.fromTag(context.stack.enchantments).forEach(shadowEntity::setAttributeFromEnchantment)
        return ActionResult.CONSUME
    }

    override fun getEnchantability(): Int = 5
    override fun isEnchantable(stack: ItemStack?): Boolean = true

    override fun canRepair(stack: ItemStack?, ingredient: ItemStack?): Boolean {
        return ingredient.let { it != null && it.item == this }
    }
}